src/plugin/contracts/base/BaseRebalanceManager.sol

```diff
@@ -0,0 +1,984 @@
+      // // console.log("totalTokensAmount: ", totalTokensAmount);
+      if (totalTokensAmount == 0) {
+        twapResult.failedToObtainTWAP = true;
+        return twapResult;
+      }
+      // uint256 totalDepositTokenMultipliedByFactor = 10000 * twapResult.totalDepositToken;
+      uint16 percentageOfDepositToken = uint16((twapResult.totalDepositToken * 10000) / totalTokensAmount);
+      twapResult.percentageOfDepositToken = percentageOfDepositToken;
+    }
+
+    // console.log('4');
+
+    uint256 depositTokenBalance = _getDepositTokenVaultBalance();
+    // console.log('depositTokenBalance: ', depositTokenBalance);
+
+    if (depositTokenBalance > 0) {
+      uint256 totalTokensAmount = twapResult.totalDepositToken + twapResult.totalPairedInDeposit;
+      // console.log('totalTokensAmount: ', totalTokensAmount);
+      // uint256 totalDepositTokenMultipliedByFactor = depositTokenBalance * 10000;
+      // че за v42 и v43, надо чекнуть первоначальный декомпайл
+      // V42 = v41 / v40 = (10000 * depositTokenBalance) / (result.totalDepositToken + result.totalPairedInDeposit)
+      twapResult.percentageOfDepositTokenUnused = uint16((depositTokenBalance * 10000) / totalTokensAmount);
+    } else {
+      // че за v44
+      // v42 = v44 = 0;
+      twapResult.percentageOfDepositTokenUnused = 0;
+    }
+  }
+
+  function _decideRebalance(TwapResult memory twapResult) internal returns (DecideStatus, State) {
+    // тут мы более глобально чекаем, получилось ли достать твап, какая ща волатильность
+    // куча куча ифов, в итоге в одном из случаев вызываем
+    // вычисления на основе twapResult И UpdateStatus, юзается вспомогательная функция _calcPercentageDiff
+    // UpdateStatus updStatus = _updateStatus(twapResult);
+    if (twapResult.failedToObtainTWAP) {
+      return (DecideStatus.FailedToObtainTWAPOrExtremeVolatility, State.Special);
+    }
+
+    uint256 fastSlowDiff = _calcPercentageDiff(twapResult.fastAvgPriceAccountingDecimals, twapResult.slowAvgPriceAccountingDecimals);
+    uint256 fastCurrentDiff = _calcPercentageDiff(twapResult.fastAvgPriceAccountingDecimals, twapResult.currentPriceAccountingDecimals);
+    // console.log('fastSlowDiff: ', fastSlowDiff);
+    // console.log('fastCurrentDiff: ', fastCurrentDiff);
+
+    bool isExtremeVolatility = fastSlowDiff >= thresholds.extremeVolatility || fastCurrentDiff >= thresholds.extremeVolatility;
+    if (!isExtremeVolatility) {
+      bool isHighVolatility = fastSlowDiff >= thresholds.highVolatility || fastCurrentDiff >= thresholds.highVolatility;
+      if (!isHighVolatility) {
+        if (
+          !((state == State.OverInventory || state == State.Normal) &&
+            lastRebalanceCurrentPrice != 0 &&
+            twapResult.percentageOfDepositToken < thresholds.underInventoryThreshold - thresholds.dtrDelta)
+        ) {
+          (bool needToRebalance, State newState) = _updateStatus(twapResult);
```

Comment:
before _updateStatus we have to check minTime between rebalances. If it's not reached then return ToSoon status